// frames.js@0.5.0/dist/chunk-LRHQB5K7.mjs
// src/getUserDataForFid.ts
import { MessageType, UserDataType, Message } from "@farcaster/core";
async function getUserDataForFid({
  fid,
  options = {}
}) {
  const optionsOrDefaults = {
    hubHttpUrl: options.hubHttpUrl ?? "https://nemes.farcaster.xyz:2281",
    hubRequestOptions: options.hubRequestOptions ?? {}
  };
  const userDataResponse = await fetch(
    `${optionsOrDefaults.hubHttpUrl}/v1/userDataByFid?fid=${fid}`,
    optionsOrDefaults.hubRequestOptions
  );
  const { messages } = await userDataResponse.json();
  if (messages && messages.length > 0) {
    const valuesByType = messages.reduce((acc, messageJson) => {
      const message = Message.fromJSON(messageJson);
      if (message.data?.type !== MessageType.USER_DATA_ADD) {
        return;
      }
      const timestamp = message.data.timestamp;
      const { type, value } = message.data.userDataBody;
      if (!acc[type]) {
        acc[type] = { value, timestamp };
      } else if (message.data.timestamp > acc[type].timestamp) {
        acc[type] = { value, timestamp };
      }
      return acc;
    }, {});
    return {
      profileImage: valuesByType[UserDataType.PFP]?.value,
      displayName: valuesByType[UserDataType.DISPLAY]?.value,
      username: valuesByType[UserDataType.USERNAME]?.value,
      bio: valuesByType[UserDataType.BIO]?.value
    };
  } else {
    return null;
  }
}

export {
  getUserDataForFid
};
