// frames.js@0.5.0/dist/validateFrame.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/validateFrame.ts
var validateFrame_exports = {};
__export(validateFrame_exports, {
  parseButtonElement: () => parseButtonElement,
  validateFrame: () => validateFrame
});
module.exports = __toCommonJS(validateFrame_exports);
var cheerio = __toESM(require("cheerio"));

// src/utils.ts
var import_core = require("@farcaster/core");
function isFrameButtonLink(frameButton) {
  return frameButton.action === "link";
}
function isFrameButtonMint(frameButton) {
  return frameButton.action === "mint";
}
function getByteLength(str) {
  return Buffer.from(str).byteLength;
}
function isValidVersion(version) {
  if (version === "vNext") {
    return true;
  }
  const pattern = /^\d{4}-\d{2}-\d{2}$/;
  if (!pattern.test(version)) {
    return false;
  }
  return true;
}

// src/getAddressForFid.ts
var import_viem = require("viem");
var import_chains = require("viem/chains");

// src/getUserDataForFid.ts
var import_core2 = require("@farcaster/core");

// src/validateFrameMessage.ts
var import_core3 = require("@farcaster/core");

// src/getFrameMessage.ts
var import_core4 = require("@farcaster/core");

// src/getTokenFromUrl.ts
function getTokenFromUrl(url) {
  const [namespace, chainId, address, tokenId] = url.split(":");
  if (!namespace || !chainId || !address) {
    throw new Error("Invalid token URL");
  }
  return {
    namespace,
    chainId: parseInt(chainId),
    address,
    tokenId: tokenId || void 0
  };
}

// src/validateFrame.ts
function validateFrame({
  htmlString,
  url
}) {
  const $ = cheerio.load(htmlString);
  let errors = null;
  function addError({ key, message }) {
    if (!errors)
      errors = {};
    if (errors.hasOwnProperty(key) && errors[key] && Array.isArray(errors[key])) {
      console.log(`Error: ${key} ${message}`);
      errors[key].push(message);
    } else {
      errors[key] = [message];
    }
  }
  const pageTitle = $("title").text();
  if (pageTitle === void 0) {
    addError({
      message: `A <title> tag is required in order for your frames to work in Warpcast`,
      key: `<title>`
    });
  }
  const version = $("meta[property='fc:frame'], meta[name='fc:frame']").attr(
    "content"
  );
  const image = $(
    "meta[property='fc:frame:image'], meta[name='fc:frame:image']"
  ).attr("content");
  const imageAspectRatio = $(
    "meta[property='fc:frame:image:aspect_ratio'], meta[name='fc:frame:image:aspect_ratio']"
  ).attr("content");
  const postUrl = $(
    "meta[property='fc:frame:post_url'], meta[name='fc:frame:post_url']"
  ).attr("content") || url;
  const inputText = $(
    "meta[property='fc:frame:input:text'], meta[name='fc:frame:input:text']"
  ).attr("content");
  const buttonLabels = [1, 2, 3, 4].flatMap(
    (el) => $(
      `meta[property='fc:frame:button:${el}'], meta[name='fc:frame:button:${el}']`
    ).map((i, elem) => parseButtonElement(elem)).filter((i, elem) => elem !== null).toArray()
  );
  const buttonActions = [1, 2, 3, 4].flatMap(
    (el) => $(
      `meta[property='fc:frame:button:${el}:action'], meta[name='fc:frame:button:${el}:action']`
    ).map((i, elem) => parseButtonElement(elem)).filter((i, elem) => elem !== null).toArray()
  );
  const buttonTargets = [1, 2, 3, 4].flatMap(
    (el) => $(
      `meta[property='fc:frame:button:${el}:target'], meta[name='fc:frame:button:${el}:target']`
    ).map((i, elem) => parseButtonElement(elem)).filter((i, elem) => elem !== null).toArray()
  );
  let buttonsValidation = [false, false, false, false];
  const buttonsWithActions = buttonLabels.map((buttonLabel) => {
    const buttonAction = buttonActions.find(
      (action2) => action2?.buttonIndex === buttonLabel?.buttonIndex
    );
    const buttonTarget = buttonTargets.find(
      (action2) => action2?.buttonIndex === buttonLabel?.buttonIndex
    );
    if (buttonsValidation[buttonLabel.buttonIndex - 1]) {
      addError({
        message: "Duplicate button",
        key: `fc:frame:button:${buttonLabel.buttonIndex}`
      });
    }
    if (![1, 2, 3, 4].includes(buttonLabel.buttonIndex)) {
      addError({
        message: "Incorrect button index (outside of 1,2,3,4)",
        key: `fc:frame:button:${buttonLabel.buttonIndex}`
      });
    } else {
      buttonsValidation[buttonLabel.buttonIndex - 1] = true;
    }
    const action = buttonAction?.content !== void 0 ? buttonAction?.content : "post";
    if (action === "link") {
      if (!buttonTarget?.content) {
        addError({
          message: "No button target, but required for action type link",
          key: `fc:frame:button:${buttonLabel.buttonIndex}`
        });
      }
      if (!(buttonTarget?.content?.startsWith("http://") || buttonTarget?.content?.startsWith("https://"))) {
        addError({
          message: "External links MUST use the https://  or http:// protocols. ",
          key: `fc:frame:button:${buttonLabel.buttonIndex}`
        });
      }
    }
    if (!buttonTarget?.content && ["link", "mint"].includes(action)) {
      addError({
        message: `Button target is required for action type ${action}`,
        key: `fc:frame:button:${buttonLabel.buttonIndex}`
      });
    }
    if (buttonTarget?.content && !buttonAction) {
      addError({
        message: "Missing button action (should be 'mint' or 'link')",
        key: `fc:frame:button:${buttonLabel.buttonIndex}`
      });
    }
    if (!["post_redirect", "post", "mint", "link", void 0].includes(
      buttonAction?.content
    )) {
      addError({
        message: "Invalid button action specified",
        key: `fc:frame:button:${buttonLabel.buttonIndex}`
      });
    }
    if (action === "mint" && buttonTarget?.content) {
      try {
        getTokenFromUrl(buttonTarget.content);
      } catch (error) {
        addError({
          message: "Invalid CAIP-10 URL",
          key: `fc:frame:button:${buttonLabel.buttonIndex}`
        });
      }
    }
    return {
      buttonIndex: buttonLabel.buttonIndex,
      label: buttonLabel.content || "",
      target: buttonTarget?.content,
      // this is an optional property, falls back to "post"
      action: buttonAction?.content || "post"
    };
  }).sort((a, b) => a.buttonIndex - b.buttonIndex).map((button) => {
    if (isFrameButtonLink(button) || isFrameButtonMint(button))
      return {
        label: button.label,
        action: button.action,
        target: button.target
      };
    return {
      label: button.label,
      action: button.action,
      target: button.target
    };
  });
  if (buttonsValidation.some((x, i) => !x && i < 3 && buttonsValidation[i + 1])) {
    addError({
      message: `Gap in buttons sequence, ${buttonsValidation.map((el, i) => `${el ? i + 1 : ""}`).join(",")}`,
      key: `fc:frame:button:1`
    });
  }
  if (!version) {
    addError({ message: "No version found in frame", key: "fc:frame" });
  } else if (!isValidVersion(version))
    addError({
      message: "Invalid version",
      key: "fc:frame"
    });
  if (!image) {
    addError({ message: "No image found in frame", key: "fc:frame:image" });
  } else if (!(image?.startsWith("http://") || image?.startsWith("https://"))) {
    if (!(image?.startsWith("data:image/png;base64,") || image?.startsWith("data:image/jpg;base64,") || image?.startsWith("data:image/jpeg;base64,") || image?.startsWith("data:image/gif;base64,"))) {
      addError({
        message: "Image has an unrecognized format. Only jpg, png and gif images are supported",
        key: "fc:frame:image"
      });
    }
  }
  if (imageAspectRatio && imageAspectRatio !== "1.91:1" && imageAspectRatio !== "1:1") {
    addError({
      message: "Invalid image aspect ratio",
      key: "fc:frame:image:aspect_ratio"
    });
  }
  if (!postUrl) {
    addError({
      message: "No post_url in frame",
      key: "fc:frame:post_url"
    });
  }
  if (getByteLength(postUrl) > 256) {
    addError({
      message: "post_url is more than 256 bytes (frames.js generates a longer post_url including system params)",
      key: "fc:frame:post_url"
    });
  }
  if (buttonsWithActions.length > 4)
    addError({ message: "Too many buttons", key: "fc:frame:button" });
  if (inputText && getByteLength(inputText) > 32) {
    addError({
      message: "Input text should be max 32 bytes",
      key: "fc:frame:input:text"
    });
  }
  return {
    frame: {
      version,
      image,
      imageAspectRatio,
      buttons: buttonsWithActions,
      postUrl,
      inputText
    },
    errors
  };
}
function parseButtonElement(elem) {
  const nameAttr = elem.attribs["name"] || elem.attribs["property"];
  const buttonIndex = nameAttr?.split(":")[3];
  try {
    return {
      buttonIndex: parseInt(buttonIndex || ""),
      content: elem.attribs["content"]
    };
  } catch (error) {
    return null;
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  parseButtonElement,
  validateFrame
});
