// frames.js@0.5.0/dist/next/client.js
"use strict";
"use client";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/next/client.tsx
var client_exports = {};
__export(client_exports, {
  FrameButtonRedirectUI: () => FrameButtonRedirectUI,
  FrameButtonUI: () => FrameButtonUI
});
module.exports = __toCommonJS(client_exports);
var import_react = __toESM(require("react"));
async function simulateAppNavigation() {
}
function FrameButtonUI(props) {
  if (typeof window === "undefined")
    return null;
  return /* @__PURE__ */ import_react.default.createElement(
    "button",
    {
      type: "button",
      onClick: () => simulateAppNavigation(),
      suppressHydrationWarning: true
    },
    props.children
  );
}
function FrameButtonRedirectUI(props) {
  if (typeof window === "undefined")
    return null;
  return /* @__PURE__ */ import_react.default.createElement(
    "button",
    {
      suppressHydrationWarning: true,
      type: "button",
      onClick: () => {
        if (props.href)
          location.href = props.href;
      }
    },
    props.children
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FrameButtonRedirectUI,
  FrameButtonUI
});
